core/vpn/src/device.rs

```diff
@@ -0,0 +1,81 @@
+use smoltcp::phy;
+use smoltcp::time;
+use std::collections::VecDeque;
+
+const MTU: usize = 65535;
+
+#[derive(Clone, Default)]
+pub struct CaptureDevice {
+    tx_queue: VecDeque<Vec<u8>>,
+    rx_queue: VecDeque<Vec<u8>>,
+}
+
+impl CaptureDevice {
+    pub fn phy_rx(&mut self, data: Vec<u8>) {
+        self.rx_queue.push_back(data);
+    }
+
+    pub fn next_phy_tx(&mut self) -> Option<Vec<u8>> {
+        self.tx_queue.pop_front()
+    }
+}
+
+impl<'a> phy::Device<'a> for CaptureDevice {
+    type RxToken = RxToken;
+    type TxToken = TxToken<'a>;
+
+    fn receive(&'a mut self) -> Option<(Self::RxToken, Self::TxToken)> {
+        let item = self.rx_queue.pop_front();
+        item.map(move |buffer| {
+            let rx = RxToken { buffer };
+            let tx = TxToken {
+                queue: &mut self.tx_queue,
+            };
+            (rx, tx)
+        })
+    }
+
+    fn transmit(&'a mut self) -> Option<Self::TxToken> {
+        Some(TxToken {
+            queue: &mut self.tx_queue,
+        })
+    }
+
+    fn capabilities(&self) -> phy::DeviceCapabilities {
+        let mut caps = phy::DeviceCapabilities::default();
+        caps.max_transmission_unit = MTU;
+        caps
+    }
+}
+
+pub struct RxToken {
+    buffer: Vec<u8>,
+}
+
+impl phy::RxToken for RxToken {
+    fn consume<R, F>(mut self, _timestamp: time::Instant, f: F) -> smoltcp::Result<R>
+    where
+        F: FnOnce(&mut [u8]) -> smoltcp::Result<R>,
+    {
+        f(&mut self.buffer)
+    }
+}
+
+pub struct TxToken<'a> {
+    queue: &'a mut VecDeque<Vec<u8>>,
+}
+
+impl<'a> phy::TxToken for TxToken<'a> {
+    fn consume<R, F>(self, _timestamp: time::Instant, len: usize, f: F) -> smoltcp::Result<R>
+    where
+        F: FnOnce(&mut [u8]) -> smoltcp::Result<R>,
+    {
+        let mut buffer = Vec::with_capacity(len);
+        buffer.resize(len, 0);
```

Comment:
this line seems to do nothing: `buffer` was just initialised with the same`len` as capacity.

I propose to:
- Remove the line
- Init the `buffer` => `with_capacity(0)`

Reply:
It fills the buffer of capacity `len` with zeroes. Before that, `buffer.len()` is equal to 0.

Playground
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7492e3da91e3adc5a7fcfc4e63c89c23

Reply:
How about this?
```rust
let mut buffer = vec![0; len];
```

Reply:
Ahhhhh, i was thinking about this option later in the evening yesterday.. TIL!

Thank you! Disregard my comment :)

I like @Wiezzel 's proposal here too, but its not as "broken" ( read sub-obtimal ) as i thought ;)